TASK-26: setup database seeding

osahtout opened the pull request:
closes #102 

Database seeding is now possible with the command npm run seed:run. This will allow to database to be populated automatically by running the command. It will: 

- assign a host for each event
- add participants to each of the events
- add friends to most users

**Steps to review PR**
1. Update your `.env` file
2. run `npm i`
3. Run the backend to update your database schema: `nx serve mull-api`
4. Run `npm run seed:run`

Note that the .env will now have to change to allow the seeding, please see the drive.

osahtout committed:
refactor (mull-api): update env variable names

Typeorm seeding needed to have specific names to connect  to the database. So now instead of DB_HOST, it's TYPEORM_HOST

refs #102

Co-authored-by: Jonathan <bit172@gmail.com>
Co-authored-by: Yunshi <dabris16.x@gmail.com>

osahtout committed:
fix (mull-api) increase length of text size

Now the description can include up to 2048 characters and the title can include update 64 characters

refs #102

Co-authored-by: Jonathan <bit172@gmail.com>
Co-authored-by: Yunshi <dabris16.x@gmail.com>

osahtout committed:
feat (mull-api): Add factories for each needed entity

For each entity, a factory is needed to seed the database with that particular entity

refs #102

Co-authored-by: Jonathan <bit172@gmail.com>
Co-authored-by: Yunshi <dabris16.x@gmail.com>

osahtout committed:
feat (mull-api): Add the seeder

The seeder will deal with creating all the entries in the database

refs #102

Co-authored-by: Jonathan <bit172@gmail.com>
Co-authored-by: Yunshi <dabris16.x@gmail.com>

osahtout committed:
chore (mull-api): cleanup

added comments, removed commented code and reduced max characters for event title

refs #102

Co-authored-by: Jonathan <bit172@gmail.com>
Co-authored-by: Yunshi <dabris16.x@gmail.com>

cristian-aldea commented on apps/mull-api/src/app/entities/event.entity.ts:
```diff
@@ -40,7 +40,7 @@ export class Event implements IEvent {
 
   @Field()
-  @Column()
+  @Column({length: '3333' })
```
Is there a reason for this specific number? should we discuss a bit about the limit we want for these columns?

osahtout replied:
We can discuss this if you want. It's just if we don't add these lengths then the maximum number of characters that the user can add is 255 which is short for a description. The seeding will also fail.

cristian-aldea replied:
Yeah sounds good, I think `3333` is a tad long, and we could a nice round number like `1000` or something

osahtout replied:
This should be a discussion though. We do have an expand feature in the mockups, maybe we can take advantages of that. 3333 is the number of characters (not the number of words) and I thought it was nice having the same as the backend url

cristian-aldea replied:
Let's discuss, again idk about putting 3333, kinda random to put it as the same number as the port of the backend app

cristian-aldea commented on apps/mull-api/src/database/seeders/database.seeder.ts:
```diff
@@ -0,0 +1,28 @@
+import { Connection } from 'typeorm';
+import { Factory, Seeder } from 'typeorm-seeding';
```
```suggestion
import { Factory, Seeder } from 'typeorm-seeding';
import { Event } from '../../app/entities';
```

running the seeder was broken without with line

osahtout replied:
fixed

cristian-aldea commented on apps/mull-api/src/database/factories/event.factory.ts:
```diff
@@ -0,0 +1,20 @@
+define(Event, (faker: typeof Faker) => {
+  const event = new Event();
+
+  event.title = faker.lorem.lines();
+  event.description = faker.lorem.paragraph();
```
You set limits for the `title` and `description` columns in `apps/mull-api/src/app/entities/event.entity.ts`. It might be wise to configure the calls to these methods so they don't go over.

osahtout replied:
We were discussing this to maybe add a limit to the front end where the words would get cut off after a certain number of characters. Again we can discuss this quickly with the team

cristian-aldea replied:
I was talking about seeding. You set a larger limit for the columns. But to make sure stuff is resilient, you should check that the `faker.lorem` calls are configured right to not go over.

Also, some suggestions after looking at the docs: https://github.com/Marak/faker.js/blob/master/doc/faker.lorem.html

you should use `faker.lorem.words` for both of these, since it seems to offer the most control (you can choose the number of words).

I'd say put 5 words for `event.title` and something like 50 for `event.description`

osahtout replied:
I don't think we should set a limit of the number of words for the description. I think we should use lorem.paragraph and give the users plenty of room to write what they want.

cristian-aldea replied:
I don't know if you understand what I'm trying to say.

The user isn't writing anything when you're seeding. I never said anything about `set a limit of the number of words for the description`. It's about making sure we don't over the limit we set for the column. I have no idea how much lorem.paragraph generates, since it's random. On the other hand, `lorem.words` is a lot easier to control. We can add more or less as we need.
![image](https://user-images.githubusercontent.com/23544999/105079461-749f7e00-5a5d-11eb-8615-32cc9c2890b2.png)
![image](https://user-images.githubusercontent.com/23544999/105079478-7ec17c80-5a5d-11eb-9209-9b556856acaf.png)

osahtout replied:
but that's the thing I don't want to control that to a certain extent. But I do want it to be random. I know the user isn't writing anything but assume that it is a user that is writing and is getting an error saying that their description is too short.

Also if it does give a text too short error, it will still seed the previous entries

osahtout replied:
I don't think the focus should be on the seeding. I think the focus should be the limit of characters in the column

cristian-aldea commented on apps/mull-api/src/environments/environment.prod.ts:
```diff
@@ -4,9 +4,9 @@ export const environment = {
-    host: process.env.DB_HOST,
-    username: process.env.DB_USERNAME,
-    password: process.env.DB_PASSWORD,
+    host: process.env.TYPEORM_HOST,
+    username: process.env.TYPEORM_USERNAME,
+    password: process.env.TYPEORM_PASSWORD,
```
Is there a reason why these were renamed? Its the db host and credentials, not the host and credentials for typeorm.

osahtout replied:
Without this, the typeorm seeding won't be able to connect the database, it was taken from [their docs](https://github.com/typeorm/typeorm/blob/master/docs/using-ormconfig.md)

cristian-aldea replied:
I see, that's fine I guess

cristian-aldea commented on package.json:
```diff
@@ -32,7 +32,9 @@
     "workspace-schematic": "nx workspace-schematic",
     "preinstall": "npx npm-force-resolutions",
-    "gen-graphql": "graphql-codegen --config codegen.yml"
+    "gen-graphql": "graphql-codegen --config codegen.yml",
```
your `package-lock.json` is weird. When I ran `npm i` it deleted 30k~ lines, please fix it by running `npm i` and then pushing the changes.

osahtout replied:
it was weird and I delete and ran npm i before pushing. i'll try again with the mac

osahtout replied:
It's still getting messed up on my end so I'll probably just commit the package.json

cristian-aldea replied:
Wait don't delete the package lock thou, I can commit a fixed version if you want while you fix your env, just ask

osahtout replied:
I'm not deleting anything. I just won't commit the package-lock.json

cristian-aldea reviewed (changes requested):
Great work! Didn't think we'd be able to solve the issues I ran into with Omar. Had a few comments though that I'd like addressed

cristian-aldea commented on apps/mull-api/src/app/entities/event.entity.ts:
```diff
@@ -40,7 +40,7 @@ export class Event implements IEvent {
 
   @Field()
-  @Column()
+  @Column({length: '3333' })
```
You didn't format this file. Please run `precommit run -a` when issuing a PR.

![image](https://user-images.githubusercontent.com/23544999/104862175-946d5f80-5900-11eb-9e82-86848b5a4792.png)

cristian-aldea commented on apps/mull-api/src/database/seeders/database.seeder.ts:
```diff
@@ -0,0 +1,28 @@
+import { Connection } from 'typeorm';
```
Please also run `nx run-many --target=lint --all` and fix the linting issues you introduced where possible: 
![image](https://user-images.githubusercontent.com/23544999/104862249-ef06bb80-5900-11eb-902e-221a38397ec0.png)

If some of the rules seems ridiculous lemme know, we could disable them so that they don't annoy us in the future

osahtout replied:
I'll try to rename some of the variables but the 'createFriends' was used before it was defined seems weird, is it because I have to have the function above it?

cristian-aldea replied:
Maybe, I'm not sure

cristian-aldea commented on package.json:
```diff
@@ -32,7 +32,9 @@
     "preinstall": "npx npm-force-resolutions",
-    "gen-graphql": "graphql-codegen --config codegen.yml"
+    "gen-graphql": "graphql-codegen --config codegen.yml",
+    "seed:config": "ts-node --project ./apps/mull-api/tsconfig.app.json -r tsconfig-paths/register ./node_modules/typeorm-seeding/dist/cli.js config",
+    "seed:run": "ts-node --project ./apps/mull-api/tsconfig.app.json -r tsconfig-paths/register ./node_modules/typeorm-seeding/dist/cli.js seed"
```
Oh Ritchelle reminded me, please add information about how to use these generator scripts in the readme

cristian-aldea commented on apps/mull-api/src/database/factories/event.factory.ts:
```diff
@@ -0,0 +1,20 @@
+
+  event.title = faker.lorem.lines();
+  event.description = faker.lorem.paragraph();
+  event.startDate = faker.date.future();
+  event.endDate = faker.date.future();
```
This won't guarantee that the date will be valid (start might be before date). `soon` might work better

Example: https://github.com/Marak/faker.js/blob/master/test/date.unit.js#L83


```suggestion
  event.startDate = faker.date.soon(5);
  event.endDate = faker.date.soon(6);
```

osahtout replied:
I believe faker.data.soon is deprecated and is no longer a function

cristian-aldea replied:
I don't see any mention of deprecation in their docs and it is definitely a function: https://github.com/Marak/faker.js/blob/master/lib/date.js#L105

You're getting that error because you're using an older version of faker, which is being used by typeorm-seeding

I changed my file to the following and the seeding worked fine. Id recommend this as a fix:
```ts
import * as faker from 'faker';
import { define, factory } from 'typeorm-seeding';
import { Event, Media, User } from '../../app/entities';

define(Event, () => {
  const event = new Event();

  event.title = faker.lorem.lines();
  event.description = faker.lorem.paragraph();
  event.startDate = faker.date.soon(3);
  event.endDate = faker.date.soon(5);
  event.restriction = faker.random.number(3);

  event.host = factory(User)() as any;
  event.coHosts = factory(User)().createMany(faker.random.number(5)) as any;
  event.participants = factory(User)().createMany(faker.random.number(30)) as any;
  event.image = factory(Media)() as any;

  return event;
});

```

osahtout replied:
faker.date.soon(number) gives :
![Screenshot 2021-01-19 151606](https://user-images.githubusercontent.com/44760251/105087976-4889fa00-5a69-11eb-9da0-0454528d8213.png)

leojrsilao reviewed (changes requested):
Good work! Noticed you have lint errors in mull-api, but overall functionality works.

osahtout committed:
feat(mull-api): add typeorm seeding packages

refs #102

osahtout committed:
chore (mull): update readme

Update the readme to include how to seed

refs #102

r-saba commented on apps/mull-api/src/database/seeders/database.seeder.ts:
```diff
@@ -0,0 +1,29 @@
+import { Event } from '../../app/entities';
+
+/**
+ * Simple algorithm that will add the next 3 users as a friend
```
No need for "Simple algorithm" just state what it does descriptively. I'm not sure what `add the next 3 users as a friend` means either without reading the code

r-saba replied:
Something like "For each user, add the next 3 users with greater id's as friends"

r-saba commented on apps/mull-api/src/database/seeders/database.seeder.ts:
```diff
@@ -0,0 +1,29 @@
+ * Simple algorithm that will add the next 3 users as a friend
+ * @param connection
+ */
+const createFriends = async (connection: Connection) => {
```
The friends table is not being populated

osahtout replied:
fixed

r-saba reviewed (changes requested):
Seeding is working great, just need to populate the friends table.

osahtout committed:
chore (mull-api): precommit, linting and improvements

improved event factory and ran linting and pre-commit

refs #102

osahtout committed:
feat (mull-api): changed seeder import and added a seed

Added a seed to have consistant data for every run. To do so, the import had to be changed.

refs #102

r-saba reviewed (approved):
LGTM! Thanks for your work

osahtout committed:
chore: update package-lock

refs #102

cristian-aldea reviewed (approved):
Everything looks great! Thanks for addressing my comments.

I can't check this right now, but make sure you've run precommit and fixed linting and tests before merging please
- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every component and function
- [x] Must match [Mockup](create-mockup-link)

osahtout commented:
> I can't check this right now, but make sure you've run precommit and fixed linting and tests before merging please

<img width="752" alt="image" src="https://user-images.githubusercontent.com/44760251/105259393-b1e93600-5b59-11eb-8835-5c79822901eb.png">

leojrsilao reviewed (approved):
Everything looks good! Linting errors are fixed and seeding works

- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every component and function
- [x] Must match [Mockup](create-mockup-link)
